### ISSUE TITLE
Can't get the same results for the cue conflict experiments

### ISSUE THREAD

**Issue by mse0** [rgeirhos/texture-vs-shape#26]
It's kind of driving me crazy because the results are pretty close and the code is pretty simple. It should be exactly the same unless torchvision changed their pretrained models. I'm not asking anyone to debug my code, but is there something obvious I'm missing here?
Resnet50: 22.2%
Alexnet: 26.3%
Vgg16: 10.7%

```
import numpy as np
import torch
from torchvision import transforms, datasets
from torch.utils.data import DataLoader
import torchvision.models as models
import probabilities_to_decision
from pathlib import Path

prob_map = probabilities_to_decision.ImageNetProbabilitiesTo16ClassesMapping()

if torch.cuda.is_available():
    print('cuda is available')
else:
    print('cuda is NOT available, using CPU')

device = torch.device("cuda" if torch.cuda.is_available() else "cpu")

batch_size = 16
img_dir = '.../texture-vs-shape/stimuli/style-transfer-preprocessed-512' 

#model = models.resnet50(pretrained=True)
#model = models.resnet18(pretrained=True)
#model = models.alexnet(pretrained=True)
model = models.vgg16(pretrained=True)
model.eval()
model.to(device)


preprocess = transforms.Compose([
    transforms.Resize(256),
    transforms.CenterCrop(224),
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406], std=[0.229, 0.224, 0.225]),
])

# imagenet categories
with open("imagenet_classes.txt", "r") as f:
    imagenet_categories = [s.strip() for s in f.readlines()]

test_data = datasets.ImageFolder(img_dir, transform=preprocess)
test_dataloader = DataLoader(test_data, batch_size=batch_size, shuffle=False)

test_data_textures = [Path(x[0]).stem.split('-')[1][:-1] for x in test_data.imgs]
test_data_textures_idx = np.array([test_data.class_to_idx[x] for x in test_data_textures])

test_data_shapes_idx = np.array(test_data.targets)

#%% Evaluate model
out = np.zeros((len(test_data), 1000))
for ii, (batch_input, _) in enumerate(test_dataloader):
    
    batch_input = batch_input.to(device)
    
    with torch.no_grad():
        tmp_out = torch.nn.functional.softmax(model(batch_input), dim=1) 
    
    out[ii*batch_size:(ii+1)*batch_size] = tmp_out.cpu().detach().numpy()

out_imagenet = [imagenet_categories[x] for x in np.argmax(out, axis=1)]
out_16 = [prob_map.probabilities_to_decision(x) for x in out]
out_16_idx = np.array([test_data.class_to_idx[x] for x in out_16])


#%% Compute shape bias
shape_correct = test_data_shapes_idx == out_16_idx 
texture_correct = test_data_textures_idx == out_16_idx 
both_correct = test_data_shapes_idx == test_data_textures_idx

# only consider correct if both are not correct
shape_correct = shape_correct & ~both_correct
texture_correct = texture_correct & ~both_correct

shape_bias = np.sum(shape_correct) / (np.sum(shape_correct) + np.sum(texture_correct))


print(f'Shape bias for {model._get_name()} is {shape_bias}')
print(f'Shape correct (only): {np.sum(shape_correct)}')
print(f'Texture correct (only): {np.sum(texture_correct)}')
print(f'Both correct: {np.sum(both_correct)}')
print(f'Neither correct: {len(test_data) - np.sum(shape_correct) - np.sum(texture_correct) - np.sum(both_correct)}')
```

**Comment by rgeirhos** [comment]
Thanks for your interest! Since the texture/shape cue conflict images already have image size 224x224, may I suggest to remove the first two steps of the preprocessing (transforms.Resize(256), transforms.CenterCrop(224))? Otherwise it will first upsample the image from 224->256 and then take a center crop. Let me know whether this does the trick.

**Comment by mse0** [comment]
I've tried changing various preprocessing steps without success. Changing the preprocessing steps can dramatically affect the shape bias though - probably mostly because it affects the number of texture decisions:

With Resize and Centercrop:
-------------------------------------
Shape bias for ResNet is 0.22208436724565755
Shape correct (only): 179
Texture correct (only): 627
Both correct: 80
Neither correct: 394

Shape bias for AlexNet is 0.2639455782312925
Shape correct (only): 194
Texture correct (only): 541
Both correct: 80
Neither correct: 465

Shape bias for VGG is 0.10749185667752444
Shape correct (only): 99
Texture correct (only): 822
Both correct: 80
Neither correct: 279

With neither Resize nor Centercrop:
-----------------------------------------------
Shape bias for ResNet is 0.2139240506329114
Shape correct (only): 169
Texture correct (only): 621
Both correct: 80
Neither correct: 410

Shape bias for AlexNet is 0.23941798941798942
Shape correct (only): 181
Texture correct (only): 575
Both correct: 80
Neither correct: 444

Shape bias for VGG is 0.07946336429308566
Shape correct (only): 77
Texture correct (only): 892
Both correct: 80
Neither correct: 231

With no Resize, Centercrop, Normalization (No preprocessing):
-----------------------------------------------------------------------------------
Shape bias for ResNet is 0.4343629343629344
Shape correct (only): 225
Texture correct (only): 293
Both correct: 80
Neither correct: 682

Shape bias for AlexNet is 0.33405172413793105
Shape correct (only): 155
Texture correct (only): 309
Both correct: 80
Neither correct: 736

Shape bias for VGG is 0.21951219512195122
Shape correct (only): 144
Texture correct (only): 512
Both correct: 80
Neither correct: 544

I was thinking maybe we did a slightly different normalization step.

**Comment by rgeirhos** [comment]
Thanks for the update. Unfortunately, I won't be able to take a closer look at this before the NeurIPS deadline (end of May); I'm afraid.

**Comment by rgeirhos** [comment]
Sorry for the extremely late reply. Not sure whether this is still relevant, but we recently open-sourced a Python toolbox which enables evaluating the shape bias of pre-trained models here: https://github.com/bethgelab/model-vs-human, perhaps this might help in resolving these issues?

**Comment by Mishalfatima** [comment]
Hi @mse0 !

Have you tried replacing the line #model = models.resnet50(pretrained=True), with 
model = models.resnet50(weights=ResNet50_Weights.IMAGENET1K_V1) for Resnet50. There are different versions of pre-trained weights, choose the one where the original accuracy matches with the one stated in this repo.
